Add documentation for user API (#570)
wazo-platform/wazo-platform.org: content/uc-doc/api_sdk/rest_api/confd/user_management.md

---
title: User Management
---

The purpose of this document is to explain how to set up a new fully operational “telephony” user (with the Wazo API), having the following features/properties:

- Voicemail enabled
- Forward enabled

> - access to Wazo Desktop enabled

Reviewer: Let's call this credentials for API usage

    ```markdown
    GET /switchboards/<switchboard_uuid>
    ```

  - Define members (existing+new user) for the switchboard (erase existing members) ([API Reference](https://wazo-platform.org/documentation/api/configuration.html#tag/switchboards/paths/~1switchboards~1%7Bswitchboard_uuid%7D/put)):
    ```markdown
    PUT /switchboards/<switchboard_uuid>/members/users
    ```

### Queue(s)

- Create a queue with user as member ([API Reference](https://wazo-platform.org/documentation/api/configuration.html#tag/queues/operation/create_queue)):
  ```markdown
  POST /queues
  ```

OR

- For each existing switchboard needed to be attached to the user:

  - Retrieve one queue with the existing members ([API Reference](https://wazo-platform.org/documentation/api/configuration.html#tag/queues/operation/get_queue)):

    ```markdown
    GET /queue/<queue_id>
    ```

  - Associate queue to the user ([API Reference](https://wazo-platform.org/documentation/api/configuration.html#tag/queues/operation/update_user_queue_association)):
    ```markdown
    PUT /queues/<queue_id>/members/users/<user_id>
    ```

> ### Func key(s)

Reviewer 2: ```suggestion
### Function Key(s)
```

  ```markdown
  PUT /users/<user_id>/funckeys
  ```

OR

- Attach a template of func keys to the user ([API Reference](https://wazo-platform.org/documentation/api/configuration.html#tag/users/operation/associate_user_func_key_template)):
  ```markdown
  PUT /users/<user_id>/funckeys/templates/<template_id>
  ```

## Create User (the new way)

As you may see in the previous chapter, there are a lot of steps to follow, in a specific order, to be able to correctly set up a new “telephony” user with all main features enabled.

To help developers to quickly create new users and easily understand the information needed, the user creation process has been recently revamped. This allows developers to create new users with just one endpoint (instead of having to call many endpoints) and to properly/automatically manage at the server-side:

- the attachments (the link between the user and their related resources)
- the rollback in case of errors (if one of the resource creations fails)

**Warning**:
The following resources must be created before:

- The switchboards
- The groups
- The funckeys templates

Below is the only step to create the user.

> ### User

Reviewer 2: ### A Single `POST` Request

- Create the “telephony” user and all the required resources needed to be fully operational ([API Reference](https://wazo-platform.org/documentation/api/configuration.html#tag/users/operation/create_user)):

>   ```markdown
>   POST /users?recursive=True
>   ```

Reviewer 2: ```suggestion
  ```markdown
  POST /users
  ```

Voilà! The user and related resources will be created from this single endpoint. If the payload is not right, nothing will be created and you will receive a complete error message.
```

---
title: User Management
---

The purpose of this page is to explain how to set up a fully operational “telephony” user (with the
Wazo API), having the following features/properties:

- Voicemail enabled
- Forward enabled
- Credentials for API usage
- One or more lines (including extensions and SIP endpoints) created
- One or more created DID numbers that allow(s) receiving calls
- Member of a ring groups
- …

Want more information about the user's API, consult the
[API documentation](/documentation/api/configuration.html#tag/users/operation/create_user).

## Create User (the new way)

As you may see in the previous chapter, there are a lot of steps to follow, in a specific order, to

> be able to correctly set up a new “telephony” user with all main features enabled.

Reviewer 2: I suggest to remove this sentence.

> To help developers to quickly create users and easily understand the information needed, the user
> creation process has been revamped. This allows developers to use one single endpoint, instead of
> having to call many endpoints. Also relations and complete validation/rollback are handled at the
> server-side:

Reviewer 3: ```suggestion
To help developers quickly create users and easily understand the information needed, the user
creation process has been revamped. This allows developers to use one single endpoint, instead of
having to call many endpoints. Also, relations and complete validation/rollback are handled on the
server-side:
```

- the attachments (the link between the user and their related resources)
- the rollback in case of errors (if one of the resource creations fails)

> **Warning**: The following resources must be created before, when needed:

Reviewer 3: ```suggestion
**Warning**: If needed, the following resources must be created before user creation:
```

- Switchboards
- Groups

> - Functions Keys Templates

Reviewer 3: The API supports creating those resources on the fly no ? 
We don't really need them to be created before the POST on `/users` from what i understood

Author: I confirm: these resources must be created before if we want them to be linked to the user. In this case, when we create a user, we provide their IDs only.

  ```markdown
  PUT /groups/<group_uuid>/members/users
  ```

### Incall(s)

- Create an incall with ‘user’ as a destination
  ([API Reference](/documentation/api/configuration.html#tag/incalls/operation/create_incall)):

  ```markdown
  POST /incalls
  ```

- Create an extension, having the context used for the incalls
  ([API Reference](/documentation/api/configuration.html#tag/extensions/operation/create_extension)):

  ```markdown
  POST /extensions
  ```

- Attach the extension(s) to the incall(s)
  ([API Reference](/documentation/api/configuration.html#tag/incalls/operation/associate_incall_extension)):
  ```markdown
  POST /incalls/<incall_id>/extensions/<extension_id>
  ```

### Service(s)

- Enable dnd and incallfilter for the user
  ([API Reference](/documentation/api/configuration.html#tag/users/operation/update_user_services)):
  ```markdown
  PUT /users/<user_id>/services
  ```

### Switchboard(s)

- Create new switchboard with user as member
  ([API Reference](/documentation/api/configuration.html#tag/switchboards/paths/~1switchboards/post)):
  ```markdown
  POST /switchboards
  ```

OR

> - For each existing switchboards needed to be attached to the user:

Reviewer 3: ```suggestion
- For each existing switchboard needed to be attached to the user:
```
